### ISSUE TITLE
modify mir-opt testing infrastructure to require that lines appear continuously

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#45153]
The mir-opt testing infrastructure ([described here](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/test/mir-opt/README.md)) allows us to write tests that do various "string matches" against the MIR output. The idea is that you can add comments at the end of [a test file like this one](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/test/mir-opt/basic_assignment.rs) that will specify output that must appear in the "dump-mir" output. [This is an example of such a comment](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/test/mir-opt/basic_assignment.rs#L38-L85). 

The basic structure is like this:

```
// END RUST SOURCE
// START rustc.node4.SimplifyCfg-initial.after.mir
//     line0
//     line1
// END rustc.node4.SimplifyCfg-initial.after.mir
```

The test infrastructure will then check that (a) `line0` appears in the given mir file and (b) `line1` appears sometime thereafter. **However,** it **also** allows for random stuff to appear in the middle. This is intended to make the test robust, but it also makes things quite surprising -- if you look at the tests, you'll see we mostly don't really want that most of the time.

What I think we should do is modify it so that `line0` and `line1` must appear **consecutively**. For bonus points, we could allow a special comment like `// ...` to be included that would indicate that zero or more lines may be skipped in between.

The code that implements this check can be found in in [`runtest.rs`](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs). It begins [here, with this loop that walks over the `.rs` test file, extracting the expected lines](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs#L2241-L2259). The expected lines are accumulated into a vector. Whenever [we reach a `// END` comment](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs#L2246), we [then invoke `compare_mir_test_output()`](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs#L2251), which will [open the relevant file](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs#L2284-L2288) and [search for the expected lines](https://github.com/rust-lang/rust/blob/72d65019c789138f555c7cf7139508d2f9f0dffe/src/tools/compiletest/src/runtest.rs#L2292).

So the idea would be to detect `// ...` and record that in the vector (e.g., we might store a `Vec<ExpectedLine>`, where `enum ExpectedLine { Ellipsis, Some(String) }`). In `compare_mir_test_output()`, we would then skip lines only when an `Ellipsis` is found.

We will possibly have to adjust some of the existing tests, I'm not sure, but we should be able to do so readily enough.

**Comment by nikomatsakis** [comment]
cc @chrisvittal, who expressed some interest in this

**Comment by chrisvittal** [comment]
This should be straightforward. I'd love to take it on.

**Comment by pnkfelix** [comment]
I believe this is resolved by #45162
